snap/hooks/configure

```diff
@@ -14,6 +83,82 @@ def configure_hook_main():
         print(f"Unsupported snap configuration: {reason}.", file=sys.stderr)
         sys.exit(1)
 
+    # Remove only base images in LXD related project
+    try:
+        lxd_images_json = subprocess.check_output(
+            ["lxc", "-f", "json", "--project", PROJECT_NAME, "image", "list"]
+        ).decode()
+    except FileNotFoundError:
+        print("LXD is not installed.", file=sys.stderr)
+        return
+    except subprocess.CalledProcessError:
+        print(f"Project {PROJECT_NAME} is not exists in the LXD.", file=sys.stderr)
+        return
+
+    lxd_images = json.loads(lxd_images_json)
+
+    for image in lxd_images:
+        for alias in image.get("aliases", []):
+            # Remove all obsolete base images
+            if any(re.fullmatch(obsolete, alias["name"]) for obsolete in OBSOLETE_VERSIONS):
+                _delete_lxd_image(image)
+                break
+
+    # Remove only base instances in LXD related project, but keep newest one per image slot.
+    try:
+        lxd_instances_json = subprocess.check_output(
+            ["lxc", "-f", "json", "--project", PROJECT_NAME, "list"]
+        ).decode()
+    except FileNotFoundError:
+        print("LXD is not installed.", file=sys.stderr)
+        return
+    except subprocess.CalledProcessError:
+        print(f"Project {PROJECT_NAME} is not exists in the LXD.", file=sys.stderr)
+        return
+
+    lxd_instances = json.loads(lxd_instances_json)
+
+    for instance in lxd_instances:
+        # Try match image.description for instance full name
+        match_result = re.fullmatch(
+            SUPPORTED_VERSIONS[0], instance["expanded_config"]["image.description"]
+        )
+        if match_result:
+            # (craft base metadata version, image provider, image os, image version)
+            image_slot = match_result.group(1, 2, 3, 4)
+            if image_slot[0] < MINIMAL_BASE_VERISON:
+                _delete_lxd_instance(instance)
+                continue
+
+            if image_slot not in image_slots:
+                # Image slot has not been recorded, save the created time and instance info
+                image_slots[image_slot] = {
+                    "time": datetime.fromisoformat(instance["created_at"][:19]),
+                    "instance": instance.copy(),
+                }
+            else:
+                # Image slot has been recorded, find and remove older version
```

Comment:
What's the problem you are attacking here?  Is there going to be multiple base instances with the exact same craft base metadata version, image provider, image os, and image version?

I think the name `image_slot` is confusing me.  Is it a dictionary of valid base instances?

Reply:
The `slot` is refer to the many possible combinations where the combination must be unique. 
`image_slot` is a dictionary of valid base instances that are unique. So it will not keep 2 versions of the same base (if that ever happened).